File: 623-add-one-row-to-tree/623-add-one-row-to-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode addOneRow(TreeNode root, int val, int depth) {
        if(depth == 1) {
            return new TreeNode(val, root, null);
        }
        
        if(depth == 2) {
            TreeNode nl = new TreeNode(val, root.left, null);
            root.left = nl;
            TreeNode nr = new TreeNode(val, null, root.right);
            root.right = nr;
        }
        
        if(root.left != null) {
            addOneRow(root.left, val, depth - 1);
        }
        
        if(root.right != null) {
            addOneRow(root.right, val, depth - 1);
        }
        
        return root;
    }
}

File: 623-add-one-row-to-tree/623-add-one-row-to-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func addOneRow(root *TreeNode, val int, depth int) *TreeNode {
    if depth == 1 {
        return &TreeNode{val, root, nil}
    }
    
    if depth == 2 {
        root.Left  = &TreeNode{val, root.Left, nil}
        root.Right = &TreeNode{val, nil, root.Right}
    }
    
    if root.Left != nil {
        addOneRow(root.Left, val, depth - 1)
    }
    
    if root.Right != nil {
        addOneRow(root.Right, val, depth - 1)
    }
    
    return root
}